// data-forge-plot/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var data_forge_1 = require("data-forge");
var data_forge_2 = require("data-forge");
var plot_api_1 = require("./plot-api");
var chart_def_1 = require("@data-forge-plot/chart-def");
exports.ChartType = chart_def_1.ChartType;
exports.AxisType = chart_def_1.AxisType;
exports.HorizontalLabelPosition = chart_def_1.HorizontalLabelPosition;
exports.VerticalLabelPosition = chart_def_1.VerticalLabelPosition;
var seriesPlotDefaults = {
    legend: {
        show: false,
    },
};
function plotSeries(plotConfig, axisMap) {
    var serializedData = this
        .inflate(function (value) { return ({ __value__: value }); })
        .serialize();
    return new plot_api_1.PlotAPI(serializedData, plotConfig || {}, axisMap || {}, seriesPlotDefaults);
}
data_forge_1.Series.prototype.startPlot = plot_api_1.startPlot;
data_forge_1.Series.prototype.endPlot = plot_api_1.endPlot;
data_forge_1.Series.prototype.plot = plotSeries;
var dataFramePlotDefaults = {
    legend: {
        show: true,
    },
};
function plotDataFrame(plotDef, axisMap) {
    var serializedData = this.serialize();
    return new plot_api_1.PlotAPI(serializedData, plotDef || {}, axisMap || {}, dataFramePlotDefaults);
}
data_forge_2.DataFrame.prototype.startPlot = plot_api_1.startPlot;
data_forge_2.DataFrame.prototype.endPlot = plot_api_1.endPlot;
data_forge_2.DataFrame.prototype.plot = plotDataFrame;
