// data-forge/build/lib/iterators/series-rolling-window-iterator.js
"use strict";
//
// Iterates an underlying iterable in the 'windows'.
//
Object.defineProperty(exports, "__esModule", { value: true });
var series_1 = require("../series");
var SeriesRollingWindowIterator = /** @class */ (function () {
    function SeriesRollingWindowIterator(iterable, period) {
        this.iterable = iterable;
        this.period = period;
    }
    SeriesRollingWindowIterator.prototype.next = function () {
        if (!this.curWindow) {
            this.curWindow = [];
            this.iterator = this.iterable[Symbol.iterator]();
            for (var i = 0; i < this.period; ++i) {
                var curPos = this.iterator.next();
                if (curPos.done) {
                    // Underlying iterator doesn't have required number of elements.
                    return { done: true };
                }
                this.curWindow.push(curPos.value);
            }
        }
        else {
            this.curWindow.shift(); // Remove first item from window.
            var curPos = this.iterator.next();
            if (curPos.done) {
                // Underlying iterator doesn't have enough elements left.
                return { done: true };
            }
            this.curWindow.push(curPos.value); // Add next item to window.
        }
        var window = new series_1.Series({
            pairs: this.curWindow
        });
        return {
            value: window,
            done: false,
        };
    };
    return SeriesRollingWindowIterator;
}());
exports.SeriesRollingWindowIterator = SeriesRollingWindowIterator;
