// @data-forge-plot/chart-def/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Defines the type of chart to output.
 */
var ChartType;
(function (ChartType) {
    ChartType["Line"] = "line";
    ChartType["Bar"] = "bar";
    ChartType["Scatter"] = "scatter";
    ChartType["Area"] = "area";
    ChartType["Histograph"] = "histogram";
    ChartType["Pie"] = "pie";
    ChartType["Donut"] = "donut";
    ChartType["RadialBar"] = "radialBar";
    ChartType["Bubble"] = "bubble";
    ChartType["Heatmap"] = "heatmap";
    ChartType["Candlestick"] = "candlestick";
    ChartType["Radar"] = "radar";
})(ChartType = exports.ChartType || (exports.ChartType = {}));
/**
 * Defines the type of an axis.
 */
var AxisType;
(function (AxisType) {
    AxisType["Numerical"] = "numerical";
    AxisType["Timeseries"] = "timeseries";
    AxisType["Category"] = "category";
})(AxisType = exports.AxisType || (exports.AxisType = {}));
/**
 * Defines the position of a horizontal label.
 */
var HorizontalLabelPosition;
(function (HorizontalLabelPosition) {
    HorizontalLabelPosition["InnerRight"] = "inner-right";
    HorizontalLabelPosition["InnerCenter"] = "inner-center";
    HorizontalLabelPosition["InnerLeft"] = "inner-left";
    HorizontalLabelPosition["OuterRight"] = "outer-right";
    HorizontalLabelPosition["OuterCenter"] = "outer-center";
    HorizontalLabelPosition["OuterLeft"] = "outer-left";
})(HorizontalLabelPosition = exports.HorizontalLabelPosition || (exports.HorizontalLabelPosition = {}));
/**
 * Defines the position of a vertical label.
 */
var VerticalLabelPosition;
(function (VerticalLabelPosition) {
    VerticalLabelPosition["InnerTop"] = "inner-top";
    VerticalLabelPosition["InnerMiddle"] = "inner-middle";
    VerticalLabelPosition["InnerBottom"] = "inner-bottom";
    VerticalLabelPosition["OuterTop"] = "outer-top";
    VerticalLabelPosition["OuterMiddle"] = "outer-middle";
    VerticalLabelPosition["OuterBottom"] = "outer-bottom";
})(VerticalLabelPosition = exports.VerticalLabelPosition || (exports.VerticalLabelPosition = {}));
